Index: mingw-ocaml/build/findlib/src/findlib/Makefile
===================================================================
--- mingw-ocaml.orig/build/findlib/src/findlib/Makefile	2013-03-25 07:57:15.116394979 -0500
+++ mingw-ocaml/build/findlib/src/findlib/Makefile	2013-03-25 07:58:44.399727226 -0500
@@ -11,8 +11,8 @@
 
 NAME = findlib
 
-OCAMLC = ocamlc
-OCAMLOPT = ocamlopt
+OCAMLC = ocamlc -I /usr/lib/ocaml/compiler-libs
+OCAMLOPT = ocamlopt -I /usr/lib/ocaml/compiler-libs
 OCAMLDEP = ocamldep
 OCAMLLEX = ocamllex
 CAMLP4O =  camlp4 pa_o.cmo pa_op.cmo pr_o.cmo --
