Index: mingw-ocaml/build/flexdll/reloc.ml
===================================================================
--- mingw-ocaml.orig/build/flexdll/reloc.ml	2013-03-22 18:18:44.440528215 -0500
+++ mingw-ocaml/build/flexdll/reloc.ml	2013-03-22 18:18:56.000000000 -0500
@@ -938,7 +938,7 @@
     parse_cmdline ();
     setup_toolchain ();
 
-    use_cygpath :=
+    use_cygpath := false; (*
       begin
         match !toolchain, !cygpath_arg with
         | _, `Yes -> true
@@ -946,7 +946,7 @@
         | (`CYGWIN|`MINGW), `None -> (Sys.command "cygpath -v 2>NUL >NUL" = 0)
         | (`MSVC|`LIGHTLD), `None -> false
       end;
-
+			  *)
 
     if !verbose >= 2 then (
       Printf.printf "** Use cygpath: %b\n" !use_cygpath;
