Index: mingw-ocaml/build/flexdll/reloc.ml
===================================================================
--- mingw-ocaml.orig/build/flexdll/reloc.ml
+++ mingw-ocaml/build/flexdll/reloc.ml
@@ -81,7 +81,9 @@ let get_output ?(use_bash = false) cmd =
       r
 
 let get_output1 ?use_bash cmd =
-  List.hd (get_output ?use_bash cmd)
+  match get_output ?use_bash cmd with
+    | x :: _ -> x
+    | [] -> ""
 
 
 (* Preparing command line *)
