Index: mingw-ocaml/build/flexdll/reloc.ml
===================================================================
--- mingw-ocaml.orig/build/flexdll/reloc.ml	2013-03-23 08:17:49.419422123 -0500
+++ mingw-ocaml/build/flexdll/reloc.ml	2013-03-23 08:19:31.578877699 -0500
@@ -82,7 +82,9 @@
       r
 
 let get_output1 ?use_bash cmd =
-  List.hd (get_output ?use_bash cmd)
+  match get_output ?use_bash cmd with
+    | x :: _ -> x
+    | [] -> ""
 
 
 (* Preparing command line *)
