Index: mingw-ocaml/build/ocaml/asmrun/i386.S
===================================================================
--- mingw-ocaml.orig/build/ocaml/asmrun/i386.S	2014-01-26 18:46:40.988419898 -0600
+++ mingw-ocaml/build/ocaml/asmrun/i386.S	2014-01-26 18:47:56.407180671 -0600
@@ -75,6 +75,9 @@
         movl %esp, %ebp; \
         call mcount; \
         popl %ebp; CFI_ADJUST(-4)
+#elif defined(SYS_mingw)
+#define PROFILE_CAML
+#define PROFILE_C
 #elif defined(SYS_bsd_elf)
 #define PROFILE_CAML \
         pushl %ebp; CFI_ADJUST(4); \
