Index: mingw-ocaml/build/ocaml/otherlibs/systhreads/Makefile.nt
===================================================================
--- mingw-ocaml.orig/build/ocaml/otherlibs/systhreads/Makefile.nt	2014-01-26 18:46:13.010363423 -0600
+++ mingw-ocaml/build/ocaml/otherlibs/systhreads/Makefile.nt	2014-01-26 18:46:12.966366479 -0600
@@ -74,6 +74,7 @@
 install:
 	cp dllthreads.dll $(STUBLIBDIR)/dllthreads.dll
 	cp libthreads.$(A) $(LIBDIR)/libthreads.$(A)
+	cd $(LIBDIR); $(RANLIB) libthreads.$(A)
 	mkdir -p $(LIBDIR)/threads
 	cp $(CMIFILES) threads.cma $(LIBDIR)/threads
 	rm -f $(LIBDIR)/threads/stdlib.cma
@@ -81,6 +82,7 @@
 
 installopt:
 	cp libthreadsnat.$(A) $(LIBDIR)/libthreadsnat.$(A)
+	cd $(LIBDIR); $(RANLIB) libthreadsnat.$(A)
 	cp $(THREAD_OBJS:.cmo=.cmx) threads.cmxa threads.$(A) $(LIBDIR)/threads
 	cp threads.cmxs $(LIBDIR)/threads
 
